diff -Naurd mpfr-4.1.1-p1-c/Makefile.in mpfr-4.1.1-p1-c2/Makefile.in
--- mpfr-4.1.1-p1-c/Makefile.in	2022-12-02 11:35:27.147596084 +0100
+++ mpfr-4.1.1-p1-c2/Makefile.in	2022-12-07 11:11:20.149645298 +0100
@@ -406,11 +406,12 @@
   examples/version.c
 
 
-# Note: Even though tarballs do not normally contain a .git directory,
-# we need to distribute version-ext.sh to avoid a failure in case a
-# tarball is tested from a Git directory (this script is executed by
-# tests/Makefile.am in case of a Git working tree), which happens with
-# "make distcheck", for instance.
+# Note: The version-ext.sh script is normally no longer needed because
+# the output_info code in tests/Makefile.am now executes this script
+# only if $(top_srcdir)/.git exists and is readable, but this should
+# not occur in a tarball (built based on EXTRA_DIST), unless someone
+# has modified it for some reason. But since this script is small, we
+# distribute it just in case it would be used.
 EXTRA_DIST = PATCHES VERSION doc/README.dev doc/check-typography doc/mini-gmp \
   tools/check_mparam.c tools/ck-clz_tab tools/ck-copyright-notice \
   tools/ck-mparam tools/ck-news tools/ck-version-info tools/get_patches.sh \
diff -Naurd mpfr-4.1.1-p1-c/tests/Makefile.in mpfr-4.1.1-p1-c2/tests/Makefile.in
--- mpfr-4.1.1-p1-c/tests/Makefile.in	2022-12-02 11:35:27.142596061 +0100
+++ mpfr-4.1.1-p1-c2/tests/Makefile.in	2022-12-07 11:12:12.549841492 +0100
@@ -1646,20 +1646,18 @@
 top_srcdir = @top_srcdir@
 AUTOMAKE_OPTIONS = 1.6 gnu
 
-# For Git, output dash-separated information:
-#   * the detected branch;
-#   * the total commit count;
-#   * the commit id (hash).
-# Note: for the branch detection, in the case of a detached HEAD state,
-# the commit may appear in multiple branches, i.e. which diverge after
-# the commit; thus we exclude branches created after this commit, based
-# on <branch>-root tags (such a tag should be added by the user when
-# creating a branch, so that "git diff <branch>-root" shows commits done
-# in the branch since its creation, etc.). If $gitb contains multiple
-# branches, this means that something is probably wrong with the tags
-# or the branches (merged branches should be deleted).
-# Concerning the test "git rev-parse ...", if git is not installed, this
-# test should silently fail.
+# For Git:
+#   * To avoid escaping the MPFR source tree, we check whether
+#     $(top_srcdir)/.git is readable (note that this is not
+#     necessarily a directory: in case of a "linked worktree",
+#     this is a file; see the git-worktree(1) man page).
+#     In particular, if one extracts an MPFR tarball within a
+#     Git working tree (e.g. after "make dist"), then running
+#     "make check" in this extracted source tree will work as
+#     expected, giving correct version information.
+#   * Additional test with "git rev-parse ...", which should
+#     silently fail if git is not installed.
+#   * See version-ext.sh for information on its output.
 output_info = { \
   if [ -d $(top_srcdir)/.svn ]; then \
     output=`svnversion $(top_srcdir) 2> /dev/null`; \
@@ -1667,7 +1665,8 @@
       echo "[svnversion] $$output"; \
     fi; \
   fi; \
-  if [ "x`git rev-parse --is-inside-work-tree 2> /dev/null`" = xtrue ]; then \
+  if [ -r $(top_srcdir)/.git ] && \
+     [ "x`git rev-parse --is-inside-work-tree 2> /dev/null`" = xtrue ]; then \
     gitv=`GREP=$(GREP) SED=$(SED) $(top_srcdir)/version-ext.sh`; \
     echo "[git] $$gitv"; \
   fi; \
