diff --git a/src/mpfr-impl.h b/src/mpfr-impl.h
index fe54cecab..613e8c1cf 100644
--- a/src/mpfr-impl.h
+++ b/src/mpfr-impl.h
@@ -1655,8 +1655,10 @@ do {                                                                  \
    this can be found in practice: https://reviews.llvm.org/D27167 says:
    "I found this problem on FreeBSD 11, where thousands_sep in fr_FR.UTF-8
    is a no-break space (U+00A0)."
-   Note, however, that this is not allowed by the C standard, which just
-   says "character" and not "multibyte character".
+   Under Linux, this is U+202F NARROW NO-BREAK SPACE (e2 80 af).
+   And in the ps_AF locale,
+     decimal_point = U+066B ARABIC DECIMAL SEPARATOR (d9 ab)
+     thousands_sep = U+066C ARABIC THOUSANDS SEPARATOR (d9 ac)
    In the mean time, in case of non-single-byte character, revert to the
    default value. */
 #if MPFR_LCONV_DPTS
diff --git a/tests/tfprintf.c b/tests/tfprintf.c
index 5af54b023..84bd45a75 100644
--- a/tests/tfprintf.c
+++ b/tests/tfprintf.c
@@ -55,7 +55,7 @@ https://www.gnu.org/licenses/ or write to the Free Software Foundation, Inc.,
 #define check_length_with_cmp(num_test, var, value, cmp, var_spec)      \
   if (cmp != 0)                                                         \
     {                                                                   \
-      mpfr_printf ("Error in test #%d, mpfr_vfprintf printed %"         \
+      mpfr_printf ("Error in test #%d: mpfr_vfprintf printed %"         \
                    QUOTE(var_spec) " characters instead of %d\n",       \
                    (num_test), (var), (value));                         \
       exit (1);                                                         \
@@ -212,7 +212,9 @@ check_mixed (FILE *fout)
   check_vfprintf (fout, "%% a. %RNg, b. %Qx, c. %td%tn", mpfr, mpq, p, &p);
   if (p != 20)
     {
-      mpfr_fprintf (stderr, "Error in test 8, got '%% a. %RNg, b. %Qx, c. %td'\n", mpfr, mpq, saved_p);
+      mpfr_fprintf (stderr,
+                    "Error in test #8: got '%% a. %RNg, b. %Qx, c. %td'\n",
+                    mpfr, mpq, saved_p);
 #if defined(__MINGW32__) || defined(__MINGW64__)
       fprintf (stderr,
                "Your MinGW may be too old, in which case compiling GMP\n"
diff --git a/tests/tprintf.c b/tests/tprintf.c
index c8a7b90de..8e4e0d2ee 100644
--- a/tests/tprintf.c
+++ b/tests/tprintf.c
@@ -62,7 +62,7 @@ https://www.gnu.org/licenses/ or write to the Free Software Foundation, Inc.,
 #define check_length_with_cmp(num_test, var, value, cmp, var_spec)      \
   if (cmp != 0)                                                         \
     {                                                                   \
-      mpfr_fprintf (stderr, "Error in test #%d, mpfr_printf printed %"  \
+      mpfr_fprintf (stderr, "Error in test #%d: mpfr_printf printed %"  \
                     QUOTE(var_spec)" characters instead of %d\n",       \
                     (num_test), (var), (value));                        \
       exit (1);                                                         \
@@ -332,7 +332,9 @@ check_mixed (void)
   check_vprintf ("%% a. %RNg, b. %Qx, c. %td%tn", mpfr, mpq, p, &p);
   if (p != 20)
     {
-      mpfr_fprintf (stderr, "Error in test 8, got '%% a. %RNg, b. %Qx, c. %td'\n", mpfr, mpq, saved_p);
+      mpfr_fprintf (stderr,
+                    "Error in test #8: got '%% a. %RNg, b. %Qx, c. %td'\n",
+                    mpfr, mpq, saved_p);
 #if defined(__MINGW32__) || defined(__MINGW64__)
       fprintf (stderr,
                "Your MinGW may be too old, in which case compiling GMP\n"
